Improvements to conversion selection.

boudewijn-tribler opened the pull request.

boudewijn-tribler committed:
Iterate over available conversions when a message is made.

Previously when a message is made the default conversion was used,
unless otherwise specified.  Now we will iterate (in reverse order)
over the available conversions and use the first (i.e. latest)
conversion that is capable of encoding the message.

boudewijn-tribler committed:
Changes _conversions from dictionary into a list.

boudewijn-tribler committed:
Asserts no longer apply.

boudewijn-tribler committed:
Cleanup for get_conversion().

boudewijn-tribler committed:
Documentation and assert cleanup.

boudewijn-tribler committed:
Clarified docstring.

NielsZeilemaker commented on community.py:
```diff
@@ -1279,51 +1280,64 @@ def get_members_from_id(self, mid):
+        When USING is a *Message* or *Message.Implementation* instance this method returns the first available
+        conversion that can *encode* USING, this is tested in reversed order using conversion.can_encode_message(USING).
+        Typically a conversion can encode a message when: the conversion knows how to encode messages with USING.name.
 
```
I get why you changed it, but "using" isn't really a proper keyword.
Maybe using two keywords is a better approach.

Then you can get_conversion(message = X) or get_conversion(packet=Y).
Add an assert which checks if only one is passed and you're done.

boudewijn-tribler replied:
I agree that splitting it up in two variables would solve the naming problem, but it feels like calling get_conversion without the need to specify the keyword argument is more user friendly.

Would a more appropriate name be better? In which case I am open to suggestions.

NielsZeilemaker replied:
I guess calling get_conversion directly isn't really something a user should do.
If he does, choosing the correct parameter name isn't really that difficult.

boudewijn-tribler replied:
An alternative is to have several methods (get_default_conversion, get_conversion_for_packet, get_conversion_for_message).  This would be in-line with other get methods already in Dispersy, such as get_member, get_member_from_id.

whirm replied:
I like last @boudewijn-tribler's proposal slightly more than the others.

boudewijn-tribler committed:
Split get_conversion(USING) into three methods.

- avoid confusion with the property USING
- use similar schema as get_member, get_member_from_id

boudewijn-tribler committed:
Import not needed outside __debug__.

NielsZeilemaker commented on community.py:
```diff
@@ -1279,51 +1279,81 @@ def get_members_from_id(self, mid):
+                return conversion
 
-    def add_conversion(self, conversion, default=False):
+        # for backwards compatibility we will raise a KeyError when no conversion for PREFIX is found (previously
```
Maybe change PREFIX with packet, otherwise I think it's a great improvement.

boudewijn-tribler committed:
Fixes typo in comment.
